### ISSUE TITLE
Regression in the termination checker

### ISSUE THREAD

**Issue by gallais** [agda/agda#4702]
This typechecks in 2.6.1 but is rejected with a termination error by current master

```agda
open import Agda.Builtin.Nat

case_of_ : {A B : Set} → A → (A → B) → B
case x of f = f x

plus : Nat → Nat → Nat
plus m n = case m of λ
   { zero    → n
   ; (suc m) → suc (plus m n)
   }
````

**Comment by jespercockx** [comment]
I guess this is related to the removal of automatic inlining (https://github.com/agda/agda/issues/4681)

**Comment by gallais** [comment]
Oh so `plus` was getting auto-inlined in the recursive call thus making the
anonymous pattern-matching lambda recursive? I did not think anonymous
lambdas could be recursive.

Edit: ah, no, my bad: `case_of_` is the one that needs inlining!

**Comment by gallais** [comment]
Is there any way we could look at the definitions used in the unhappy path
and advise users to introduce `{-# INLINE #-}` pragmas as a potential fix?

**Comment by jespercockx** [comment]
No, I think it is the function `case_of_` that was getting auto-inlined. So you'd need a pragma `{-# INLINE case_of_ #-}`.

**Comment by UlfNorell** [comment]
I suppose we could check if there are any problematic calls to functions that satisfy the auto-inline criterion...

**Comment by mwageringel** [comment]
Here is another example for which termination checking fails. This fails with Agda 2.6.1, but used to work with earlier versions. I have not been able to try the current master.

```agda
open import Agda.Builtin.Bool
open import Agda.Builtin.List

module _ {α} {A : Set α} (_≤?_ : A → A → Bool) where

  merge : List A → List A → List A
  merge xs@(x ∷ xs′) ys@(y ∷ ys′) with x ≤? y
  ... | true = x ∷ merge xs′ ys
  ... | false = y ∷ merge xs ys′
  merge [] ys = ys
  merge xs [] = xs
```

**Comment by gallais** [comment]
@mwageringel This is related to `with`-inlining and not
to inlining of small functions. Cf #3604 and #3794 for
the issue & PR, and [the doc](https://agda.readthedocs.io/en/v2.6.1/language/with-abstraction.html#termination-checking) for ways of fixing your code.

**Comment by mwageringel** [comment]
@gallais Thank you so much for the explanation and the links, and sorry for the noise.

**Comment by andreasabel** [comment]
Workaround: `{-# INLINE case_of #-}`

Doing this automatically would require some substantial changes to the termination checker. (A auto-inline pass.)

Looks like we will release this regression in 2.6.2.
TODO: document in release notes!

**Comment by andreasabel** [comment]
Documented the regression with the OP as example.

Putting it into the icebox, we might want to add auto-inlining for the sake of termination checking.
